qutip/core/cy/coefficient.pyx

```diff
@@ -16,6 +17,62 @@ cdef extern from "<complex>" namespace "std" nogil:
     double complex conj(double complex x)
     double         norm(double complex x)
 
+
+def coefficient_function_parameters(func, style=None):
+    """
+    Return the function style (either "pythonic" or not) and a list of
+    additional parameters accepted.
+
+    Used by :obj:`FunctionCoefficient` and :obj:`_FuncElement` to determine the
+    call signature of the supplied function based on the given style (or
+    ``qutip.settings.core["function_coefficient_style"]`` if no style is given)
+    and the signature of the given function.
+
+    Parameters
+    ----------
+    func : function
+        The :obj:`FunctionCoefficient` to inspect. The first argument
+        of the function is assumed to be ``t`` (the time at which to
+        evaluate the coefficient). The remaining arguments depend on
+        the signature style (see below).
+
+    style : {None, "pythonic", "dict", "auto"}
+        The style of the signature used. If style is ``None``,
+        the value of ``qutip.settings.core["function_coefficient_style"]``
+        is used. Otherwise the supplied value overrides the global setting.
+
+    Returns
+    -------
+    (f_pythonic, f_parameters)
+
+    f_pythonic : bool
+        True if the function should be called as ``f(t, **kw)`` and False
+        if the function should be called as ``f(t, kw_dict)``.
+
+    f_parameters : set or None
+        The set of parameters (other than ``t``) of the function or
+        ``None`` if the function accepts arbitrary parameters.
+    """
+    sig = inspect.signature(func)
+    f_has_kw = any(p.kind == p.VAR_KEYWORD for p in sig.parameters.values())
+    if style is None:
+        style = qutip.settings.core["function_coefficient_style"]
+    if style == "auto":
+        if tuple(sig.parameters.keys()) == ("t", "args") and not f_has_kw:
+            # if the signature is exactly f(t, args), then assume parameters
+            # are supplied in an argument dictionary
+            style = "dict"
```

Comment:
We never said to the user to use this exact signature and adding it now calling it _backward compatibility_ feels wrong. We can do better.
_Function that if called with `f(t, **args)` would be missing the second positional argument._ would be a good condition.
These functions would fail if called as pythonic so we are still giving it more weight, it's more inclusive than a strict signature without being overly greedy. This would support `f(t, _)` and `f(t, kwargs)` which are other natural signature for the dict format. My expectation is that for most user not using decorator, this would work as expected. Users using `partial` etc. will need to update their code, but the `function_style` is there for them.
 
```suggestion
        keys = tuple(sig.parameters.keys())
        if (
            len(keys) == 2
            and not f_has_kw
            and keys[1] not in args
            and sig.parameters[keys[1]].default is inspect._empty
        ):
```

Reply:
My understanding was there would be no `style='auto'`, just `'dict'` or `'pythonic'`. So I saw `'auto'` as a try at backward compatibility, for which it's not great, but as a way to support both v4 and v5 by projects using qutip, it's look fine. These projects would use `f(t, args)` to support both version until they start using other new features and drop v4, but that is still far ahead.

`qutip.settings.core`'s options will be useful for users trying out v5. My guess is that many users will try both version for their research and take the fastest if they don't need the new features. (openmp will probably not be working on v5 in the beginning, etc.) So I want as little friction in passing from v4 to v5 for normal qutip usage.

About point 4, we should add context with options, while not needed here with this solution, it would be nice for testing global options if nothing else. But it would allow project to use their own options if needed.